### PR TITLE
Read History

### PR DESCRIPTION
**Pull request by chimtrangbu**

### CODE REVIEW

File: components/Header/ReadHistory.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import './ReadHistory.less';
+import {useSelector} from "react-redux";
+
+const ReadHistory = (props) => {
+  const [readHistory, setReadHistory] = useState([])
+  const { isLoggedIn } = useSelector((state) => {
+    return state.users;
+  });
+
+  useEffect(async () => {
```
Comment (HellCatVN) [inline components/Header/ReadHistory.js:14]: No async useEffect function

File: components/Header/ReadHistory.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            </Col>
+          )): <div>Chưa có lịch sử đọc truyện</div>}
+        </React.Fragment>
+      </dd>
+    </div>
+  )
+}
+
+// TODO: get cookies to pass basic middleware
+export async function getServerSideProps({res, params, req}) {  // didn't work
```
Comment (HellCatVN) [inline components/Header/ReadHistory.js:70]: Remove ?

File: pages/truyen/[[...slug]].js [hunk 3]
```diff
@@ -37,17 +38,29 @@ const Comic = ({ chapterData }) => {
...
+    console.log('readHistory', readHistory)
+    localStorage.setItem("readHistory", JSON.stringify(readHistory))
 
     if(isLoggedIn){
-
+      const saveDbResult = await markReadAPI({
+        comicId: chapterData.chapter_info.comicId._id,
+        chapterId:  chapterData.chapter_info._id,
+      })
+      if (saveDbResult && saveDbResult.data.code === 200) {
```
Comment (HellCatVN) [inline pages/truyen/[[...slug]].js:58]: when markAsRead Push to localStorage and update it no need to get from API again to fetch new data

File: components/Header/ReadHistory.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+const ReadHistory = (props) => {
+  const [readHistory, setReadHistory] = useState([])
+  const { isLoggedIn } = useSelector((state) => {
+    return state.users;
+  });
+
+  useEffect(async () => {
+    console.log('props', props) // undefined :(
+    const localReadHistory = JSON.parse(localStorage.getItem("readHistory")) || [];
+    if (isLoggedIn) {
```
Comment (HellCatVN) [inline components/Header/ReadHistory.js:17]: This is will cause multiple times call getAPI if we render this component or update page

Comment (HellCatVN) [reply components/Header/ReadHistory.js:17]: use Like bz

File: pages/lich-su-doc-truyen/index.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    return state.users;
+  });
+
+  useEffect(() => {
+    const localReadHistory = JSON.parse(localStorage.getItem("readHistory"));
+    if (!localReadHistory && isLoggedIn) {
+      getReadAPI(
+        ["title", "thumbnail", "slug", "excerpt", "author", "views"],
+        ["chapterIndex", "slug", "fakeSlug"]
+      ).then(result => {
```
Comment (HellCatVN) [inline pages/lich-su-doc-truyen/index.js:27]: We don't pass the field to backend. backend can select it don't let user decide what they need

File: components/Header/ReadHistory.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+const ReadHistory = () => {
+  const [readHistory, setReadHistory] = useState([])
+  const { isLoggedIn } = useSelector((state) => {
+    return state.users;
+  });
+
+  useEffect(() => {
+    const localReadHistory = JSON.parse(localStorage.getItem("readHistory"));
+    if (!localReadHistory && isLoggedIn) {
```
Comment (HellCatVN) [inline components/Header/ReadHistory.js:16]: Where to check first time load sync history ?

Comment (chimtrangbu) [reply components/Header/ReadHistory.js:16]: localReadHistory is undefined means that history has never sync. If localReadHistory has some value we'll pass the fetchApi step.

Comment (HellCatVN) [reply components/Header/ReadHistory.js:16]: So if user read new comic on other devices how can we sync data ?

**Pull request merged**